### ISSUE TITLE
Issue with default service account in Google Cloud Run

### ISSUE THREAD

**Issue by jrmerz** [googleapis/google-auth-library-nodejs#800]
I am having trouble connecting to Google Pub/Sub in Google Cloud Run using the default service account.  I am seeing the following in the Google Cloud Run service logs:

```
(node:14) UnhandledPromiseRejectionWarning: Error: Could not load the default credentials. Browse to https://cloud.google.com/docs/authentication/getting-started for more information.
GoogleAuth.getApplicationDefaultAsync (/usr/src/app/node_modules/google-auth-library/build/src/auth/googleauth.js:161:19)
processTicksAndRejections (internal/process/task_queues.js:93:5)
runNextTicks (internal/process/task_queues.js:62:3)
listOnTimeout (internal/timers.js:501:9)
processTimers (internal/timers.js:475:7)
async GoogleAuth.getClient (/usr/src/app/node_modules/google-auth-library/build/src/auth/googleauth.js:503:17)
async GrpcClient._getCredentials (/usr/src/app/node_modules/google-gax/build/src/grpc.js:108:24)
async GrpcClient.createStub (/usr/src/app/node_modules/google-gax/build/src/grpc.js:229:23)
```

I have setup a sample using a modified version of the quick start tutorial for Google Cloud Run to replicate issue: https://github.com/ucd-library/gcr-default-service-account.

Is there something I'm missing (https://github.com/ucd-library/gcr-default-service-account/blob/master/index.js#L6) or is this a bug?  Seems like it could be similar to #320?

#### Environment details

  - OS: Linux.  Google Cloud Run container.
  - Node.js version: 12
  - npm version:
  - `google-auth-library` version: 5.3.0 (version used by *current* latest ```@google-cloud/pubsub@1.1.0```)

#### Steps to reproduce

  1. Deploy container to Google Cloud Run
  2. View error in Cloud Run console

**Comment by savil** [comment]
I'm running into a similar issue ("could not load the default credentials") trying to access Datastore from a container on Cloud Run.

A minimal repro repository is at: https://github.com/savil/gcloud-run-credentials-datastore/blob/master/app.js

```
2019-10-16T19:13:48.676941Z something went wrong Error: Could not load the default credentials. Browse to https://cloud.google.com/docs/authentication/getting-started for more information. A 
2019-10-16T19:13:48.676986Z     at GoogleAuth.getApplicationDefaultAsync (/use/src/app/node_modules/google-auth-library/build/src/auth/googleauth.js:161:19) A 
2019-10-16T19:13:48.677054Z     at process._tickCallback (internal/process/next_tick.js:68:7)  

```

OS: Cloud run container
Node js: v10.16.3
`"@google-cloud/datastore": "4.4.0"`
`"google-auth-library": "5.5.0"`

**Comment by bcoe** [comment]
@savil does this problem happen consistently, or is it occurring periodically?

**Comment by savil** [comment]
@bcoe seems to happen consistently for me

**Comment by savil** [comment]
are you able to repro if you run the code in https://github.com/savil/gcloud-run-credentials-datastore/ ?

**Comment by bcoe** [comment]
@savil first of all, I would like to provide a potential workaround:

You can create a service account, rather than relying on the default credentials:

https://cloud.google.com/docs/authentication/getting-started

point_up this requires that the credentials are available in a file on disk, so you would need to upload this JSON file with your project (keeping in mind it contains secrets and shouldn't be stored anywhere insecure).

At which point you would set your credentials to GOOGLE_APPLICATION_CREDENTIALS=./my-service-account.json.

---

To help us get to the bottom of the underlying issue, could I bother you to run your application with `DEBUG_AUTH = true`, I believe the issue might be similar to #798; where the metadata endpoint we use to determine the default credentials is timing out.

@jrmerz are you also continuing to run into this issue?

**Comment by jrmerz** [comment]
I have redeployed, looks like the container failed twice and loaded the third time (```topics loaded 1``` ... so the container accessed to pub/sub).  See the last failure and final success below.  Note, I did deploy with ```--set-env-vars=DEBUG_AUTH=true``` included in the command.  Was that correct?

```
2019-10-02 10:03:30.669 PDT(node:14) UnhandledPromiseRejectionWarning: Error: Could not load the default credentials. Browse to https://cloud.google.com/docs/authentication/getting-started for more information.
2019-10-02 10:03:30.669 PDT at GoogleAuth.getApplicationDefaultAsync (/usr/src/app/node_modules/google-auth-library/build/src/auth/googleauth.js:161:19)
2019-10-02 10:03:30.669 PDT at processTicksAndRejections (internal/process/task_queues.js:93:5)
2019-10-02 10:03:30.669 PDT at runNextTicks (internal/process/task_queues.js:62:3)
2019-10-02 10:03:30.669 PDT at listOnTimeout (internal/timers.js:501:9)
2019-10-02 10:03:30.669 PDT at processTimers (internal/timers.js:475:7)
2019-10-02 10:03:30.669 PDT at async GoogleAuth.getClient (/usr/src/app/node_modules/google-auth-library/build/src/auth/googleauth.js:503:17)
2019-10-02 10:03:30.669 PDT at async GrpcClient._getCredentials (/usr/src/app/node_modules/google-gax/build/src/grpc.js:108:24)
2019-10-02 10:03:30.669 PDT at async GrpcClient.createStub (/usr/src/app/node_modules/google-gax/build/src/grpc.js:229:23)
2019-10-02 10:03:30.669 PDT(node:14) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)
2019-10-02 10:03:30.669 PDT(node:14) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
2019-10-18 09:30:47.285 PDTCloud RunDeleteServicetest-default-service-accountjrmerz@ucdavis.edu {@type: type.googleapis.com/google.cloud.audit.AuditLog, authenticationInfo: {…}, authorizationInfo: […], methodName: google.cloud.run.v1.Services.DeleteService, request: {…}, requestMetadata: {…}, resourceLocation: {…}, resourceName: namespaces/digital-ucdavis-edu/services/test-default-service-acco…
2019-10-21 19:23:49.206 PDTCloud RunCreateServicetest-default-service-accountjrmerz@ucdavis.edu {@type: type.googleapis.com/google.cloud.audit.AuditLog, authenticationInfo: {…}, authorizationInfo: […], methodName: google.cloud.serverless.v1alpha1.Services.CreateService, request: {…}, requestMetadata: {…}, resourceLocation: {…}, resourceName: namespaces/digital-ucdavis-edu/services/test-default…
2019-10-21 19:24:00.143 PDT
2019-10-21 19:24:00.143 PDT> gcr-default-service-account@1.0.0 start /usr/src/app
2019-10-21 19:24:00.143 PDT> node index.js
2019-10-21 19:24:00.143 PDT
2019-10-21 19:24:02.739 PDTHello world listening on port 8080
2019-10-21 19:24:22.952 PDTtopics loaded 1
```

Does my sample repo work for you @bcoe?  

We are not continuing to run into the issue as I could not recommend using Google Cloud Run for what we were designing with such basic functionality broken.  That said I am happy to retest and help fix.

**Comment by bcoe** [comment]
@jrmerz thank you for all the info :+1:; to answer your question, I've only been able to intermittently reproduce issues (I'm continuing to dig).

@savil do you have an example of the code that you're running?

**Comment by bcoe** [comment]
@jrmerz had a conversation with the folks on the Cloud Run team, and I think this is potentially things working as expected (although this confused be a bit too).

The cloud run processes are designed for serving HTTP traffic, and do not have their resources fully allocated unless they are currently serving traffic.

Because your requests to PubSub happen outside of a request, they cloud run container is not fully allocated, and is running at a crawl; we then fail to successfully interact with the meta-data server, and misinterpret this as failing to load credentials.

tldr; work background processes, like consuming from a queue, Cloud Run isn't the best bet. A product like GKE, or wiring up a PubSub queue to hit an HTTP endpoint, might be the better bet.

**Comment by savil** [comment]
@bcoe a minimal repro for my code would be: https://github.com/savil/gcloud-run-credentials-datastore/ 

didn't yet run with `DEBUG_AUTH=true`

> tldr; work background processes, like consuming from a queue, Cloud Run isn't the best bet. A product like GKE, or wiring up a PubSub queue to hit an HTTP endpoint, might be the better bet.

ah, yeah, my task was more of a server-side only processing task. Its not serving an external HTTP request, so this probably explains it.

**Comment by bcoe** [comment]
@savil I'm passing along feedback that this created confusion for various folks, including me :smile: I think it would be good for us to make sure we describe well the types of workflows our different compute options are good for :+1: 

Sorry this tripped you both  up.
